### PR TITLE
rsync: fix permissions and ownership issues

### PR DESCRIPTION
**Pull request by nrauso**
NethServer/dev#6691

### CODE REVIEW

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
+
+#
+# Copyright (C) 2013 Nethesis S.r.l.
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

**Review by gsanchietti — COMMENTED** [review]
I see a couple of problems with the proposed fix.

If the system has been restored without package reinstall, all related chown/chgrp command will fail because the user doesn't still exist. A later package install, will not fix the permission problem because the `fix-rsync-permission` action will not be called.

Due to the modularity of NethServer, every package should handle its own data.
The proposed solution tries to bake everything inside a script and this is quite against current conventions.

The bug is real, but probably a well-designed fix requires a lot of rework on many packages.
I strongly believe we should avoid such changes right now.
Maybe we can try to implement this partial fix but correctly split inside all involved packages.

Finally, we should also add a note inside the manual.

See also [this old implementation](https://gist.githubusercontent.com/nrauso/551e8d8617fd7b19584aff32001d0b7f/raw/e983e2b976ae07bd0874d51a71471001d7a1c9ee/fixperm) from @nrauso

**Comment by gsanchietti** [comment]
Let's wait @DavidePrincipi comment (after summer holidays).

**Comment by DavidePrincipi** [comment]
> Due to the modularity of NethServer, every package should handle its own data.
> The proposed solution tries to bake everything inside a script and this is quite against current conventions.

I agree to split the fix in affected modules. It's easier to develop/test on different products.

A test case with multiple modules is needed though, to reproduce the failure scenario.

**Comment by gsanchietti** [comment]
Ok, so let's split this PR inside multiple packages.

Each package should contain a new action linked inside the `post-restore-data` event.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+shift
+
+# check if engine used is rsync
+
+backup=$1
+[ "$(/sbin/e-smith/db backups gettype "${backup}")" = rsync ] || exit 0
+
+# if rsync then fix permissions
+
+if [[ -d "/var/lib/asterisk/" && $(stat -c %U:%G "/var/lib/asterisk/") != "asterisk:asterisk" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:34]: Move to nethserver-freepbx package.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown asterisk:asterisk /var/lib/asterisk
+  /bin/chown -R asterisk:asterisk /var/lib/asterisk/moh
+  /bin/chown -R asterisk:asterisk /var/lib/asterisk/sounds
+  /bin/chown -R asterisk:asterisk /var/spool/asterisk
+fi
+
+if [[ -d "/var/lib/nethserver/nethcti/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:41]: Move to nethcti3 package (enterprise only).

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R asterisk:asterisk /var/lib/nethserver/nethcti/templates/customer_card
+  /bin/chown -R asterisk:asterisk /var/lib/nethserver/nethcti/static
+fi
+
+if [[ -d "/var/lib/nethserver/nethvoice/phonebook/uploads/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:46]: Move to nethvoice-wizard-restapi package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chgrp -R vmail /var/lib/nethserver/imapsync
+fi
+
+if [[ -d "/var/lib/rspamd/" ]]; then
+  /bin/chown -R '_rspamd:_rspamd' /var/lib/rspamd
+  /bin/chown -R redis:redis /var/lib/redis
+  /bin/chown redis:mail /var/lib/redis/rspamd/dump.rdb
+fi
+
+if [[ -d "/var/lib/evebox/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:67]: Move to nethserver-evebox package.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R suricata:suricata /var/lib/evebox
+fi
+
+if [[ -d "/var/lib/nethserver/webtop/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:71]: Move to nethserver-webtop5 package.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R tomcat:tomcat /var/lib/nethserver/webtop
+  /bin/chown -R postgres:postgres /var/lib/nethserver/webtop/backup
+fi
+
+if [[ -d "/var/lib/nethserver/mattermost/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:76]: Move to nethserver-mattermost package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R mattermost:mattermost /opt/mattermost
+  /bin/chown -R postgres:postgres /var/lib/nethserver/mattermost/backup
+  /bin/chown -R mattermost:mattermost /var/lib/nethserver/mattermost/data
+fi
+
+if [[ -d "/var/lib/nethserver/ejabberd/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:82]: Move to nethserver-ejabberd package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R ejabberd:ejabberd /var/lib/nethserver/ejabberd
+fi
+
+# if logs are included in the backup then fix their permissions too
+
+if [ "$(/sbin/e-smith/config getprop backup-data IncludeLogs)" == enabled ]; then
+  if [[ -d "/var/log/asterisk" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:89]: Move to nethserver-freepbx package.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R asterisk:asterisk /var/log/asterisk
+  fi
+  
+  if [[ -d "/var/log/c-icap/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:93]: Move to nethserver-c-icap package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown c-icap:c-icap /var/log/c-icap
+  fi
+  
+  if [[ -d "/var/log/clamav-unofficial-sigs/" ]]; then
+    /bin/chown -R clamupdate:clamupdate /var/log/clamav-unofficial-sigs
+  fi
+  
+  if [[ -d "/var/log/ejabberd/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:101]: Move to nethserver-ejabberd package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  
+  if [[ -d "/var/log/imapsync/" ]]; then
+    /bin/chown -R vmail:vmail /var/log/imapsync
+  fi
+  
+  if [[ -d "/var/log/pbx/" ]]; then
+    /bin/chown -R asterisk:asterisk /var/log/pbx
+  fi
+  
+  if [[ -d "/var/log/redis/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:117]: This should be split between nethserver-nethvoice-report (enterprise only) and nethserver-mail-filter.

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R redis:redis /var/log/redis
+    /bin/chown redis:asterisk /var/log/redis/nethvoice-report.log*
+    /bin/chown redis:mail /var/log/redis/rspamd.log*
+  fi
+  
+  if [[ -d "/var/log/rspamd/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:123]: Move to nethserver-mail-filter

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R vmail:vmail /var/lib/nethserver/vmail
+  /bin/chgrp vmail /var/lib/nethserver/sieve-scripts
+  /bin/chown vmail:vmail /var/lib/nethserver/sieve-scripts/*.svbin
+fi
+
+if [[ -d "/var/lib/nethserver/imapsync/" ]]; then
+  /bin/chgrp -R vmail /var/lib/nethserver/imapsync
+fi
+
+if [[ -d "/var/lib/rspamd/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:61]: Move to nethserver-mail-filter

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R asterisk:asterisk /var/lib/nethserver/nethcti/templates/customer_card
+  /bin/chown -R asterisk:asterisk /var/lib/nethserver/nethcti/static
+fi
+
+if [[ -d "/var/lib/nethserver/nethvoice/phonebook/uploads/" ]]; then
+  /bin/chgrp asterisk /var/lib/nethserver/nethvoice/phonebook/uploads
+  /bin/chown -R asterisk:asterisk /var/lib/nethserver/nethvoice/phonebook/uploads/*
+fi
+
+if [[ -d "/var/lib/nethserver/vmail/" && $(stat -c %U:%G "/var/lib/nethserver/vmail/") != "vmail:vmail" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:51]: Move to nethserver-mail-server package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+  /bin/chown -R vmail:vmail /var/lib/nethserver/vmail
+  /bin/chgrp vmail /var/lib/nethserver/sieve-scripts
+  /bin/chown vmail:vmail /var/lib/nethserver/sieve-scripts/*.svbin
+fi
+
+if [[ -d "/var/lib/nethserver/imapsync/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:57]: Move to nethserver-mail-imapsync package

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+if [ "$(/sbin/e-smith/config getprop backup-data IncludeLogs)" == enabled ]; then
+  if [[ -d "/var/log/asterisk" ]]; then
+    /bin/chown -R asterisk:asterisk /var/log/asterisk
+  fi
+  
+  if [[ -d "/var/log/c-icap/" ]]; then
+    /bin/chown c-icap:c-icap /var/log/c-icap
+  fi
+  
+  if [[ -d "/var/log/clamav-unofficial-sigs/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:97]: Move to nethserver-antivirus

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+  
+  if [[ -d "/var/log/ejabberd/" ]]; then
+    /bin/chown -R ejabberd:ejabberd /var/log/ejabberd
+  fi
+  
+  if [[ -f "/var/log/freshclam.log" ]]; then
+    /bin/chown -R root:clamupdate /var/log/freshclam.log
+  fi
+  
+  if [[ -d "/var/log/imapsync/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:109]: Move to nethserver-mail-imapsync

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R redis:redis /var/log/redis
+    /bin/chown redis:asterisk /var/log/redis/nethvoice-report.log*
+    /bin/chown redis:mail /var/log/redis/rspamd.log*
+  fi
+  
+  if [[ -d "/var/log/rspamd/" ]]; then
+    /bin/chown -R _rspamd:_rspamd /var/log/rspamd
+  fi
+  
+  if [[ -d "/var/log/squid/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:127]: Move to nethserver-squid

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R squid:squid /var/log/squid
+  fi
+  
+  if [[ -d "/var/log/suricata/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:131]: Move to nethserver-suricata

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R suricata:suricata /var/log/suricata
+  fi
+  
+  if [[ -d "/var/log/ufdbguard/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:135]: Move to nethserver-squidguard

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+  
+  if [[ -d "/var/log/clamav-unofficial-sigs/" ]]; then
+    /bin/chown -R clamupdate:clamupdate /var/log/clamav-unofficial-sigs
+  fi
+  
+  if [[ -d "/var/log/ejabberd/" ]]; then
+    /bin/chown -R ejabberd:ejabberd /var/log/ejabberd
+  fi
+  
+  if [[ -f "/var/log/freshclam.log" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:105]: Move to nethserver-antivirus

File: root/etc/e-smith/events/actions/fix-rsync-permissions [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+    /bin/chown -R root:clamupdate /var/log/freshclam.log
+  fi
+  
+  if [[ -d "/var/log/imapsync/" ]]; then
+    /bin/chown -R vmail:vmail /var/log/imapsync
+  fi
+  
+  if [[ -d "/var/log/pbx/" ]]; then
```
Comment (gsanchietti) [inline root/etc/e-smith/events/actions/fix-rsync-permissions:113]: Move to nethserver-freepbx

**Review by gsanchietti — CHANGES_REQUESTED** [review]

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 2 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:34]: Moved to [nethserver-freepbx#77](https://github.com/NethServer/nethserver-freepbx/pull/77)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 9 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:89]: Moved to [nethserver-freepbx#77](https://github.com/NethServer/nethserver-freepbx/pull/77)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 23 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:113]: Moved to [nethserver-freepbx#77](https://github.com/NethServer/nethserver-freepbx/pull/77)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 4 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:46]: Moved to [nethvoice-wizard-restapi#202](https://github.com/nethesis/nethvoice-wizard-restapi/pull/202)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 12 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:117]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261) and to [nethvoice-report#175](https://github.com/nethesis/nethvoice-report/pull/175)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 13 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:123]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 14 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:61]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 15 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:51]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 16 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:57]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 18 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:109]: Moved to [nethserver-mail#261](https://github.com/NethServer/nethserver-mail/pull/261)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 20 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:131]: Moved to [nethserver-suricata#47](https://github.com/NethServer/nethserver-suricata/pull/47)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 5 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:67]: Moved to [nethserver-evebox#13](https://github.com/NethServer/nethserver-evebox/pull/13)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 6 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:71]: Moved to [nethserver-webtop5#127](https://github.com/NethServer/nethserver-webtop5/pull/127)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 7 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:76]: Moved to [nethserver-mattermost#115](https://github.com/NethServer/nethserver-mattermost/pull/115)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 8 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:82]: Moved to [nethserver-ejabberd#43](https://github.com/NethServer/nethserver-ejabberd/pull/43)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 11 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:101]: Moved to [nethserver-ejabberd#43](https://github.com/NethServer/nethserver-ejabberd/pull/43)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 17 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:97]: Moved to [nethserver-antivirus#16](https://github.com/NethServer/nethserver-antivirus/pull/16)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 22 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:105]: Moved to [nethserver-antivirus#16](https://github.com/NethServer/nethserver-antivirus/pull/16)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 19 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:127]: Moved to [nethserver-squid#78](https://github.com/NethServer/nethserver-squid/pull/78)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 21 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:135]: Moved to [nethserver-squidguard#45](https://github.com/NethServer/nethserver-squidguard/pull/45)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 10 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:93]: Moved to [nethserver-c-icap#2](https://github.com/NethServer/nethserver-c-icap/pull/2)

File: root/etc/e-smith/events/actions/fix-rsync-permissions (hunk 3 above)
Comment (nrauso) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:41]: I cannot see the repo, could you please check why?

Comment (gsanchietti) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:41]: I will do it

Comment (gsanchietti) [reply root/etc/e-smith/events/actions/fix-rsync-permissions:41]: See nethesis/nethcti-server#278

**Comment by gsanchietti** [comment]
The PR has been split between involved packages.

**Pull request closed without merging**